### DISCUSSION TITLE
Does supabase sanitize data before putting into database?

### DISCUSSION THREAD

**Discussion by hashiromer** [supabase/supabase#1452] [Questions] [answered]
A lot of people including me are trying to use Supabase without using serverless functions by relying upon PostgreSQL's
row level security. However, if our application allows users to post new data which would later be accessed by other people, we need to sanitize requests on the server before putting into our database. I would like to know if Supabase does it for us, or is it something we have to take care of ourselves by using a proxy of some sort?

**Comment by steve-chavez** [comment, marked as answer]
> we need to sanitize requests on the server before putting into our database

**SQLi**: If you use the PostgREST API(through `supabase-js` or other supabase client libs), all the inputs to the db are parametrized and the statements are prepared. No risk of a sql injection.

**XSS**: If you mean storing html values on the db, and prevent XSS when rendering them on the clients. Then this can be done client side(for example, [ReactJS escapes strings automatically](https://stackoverflow.com/a/51852579/4692662)) or if you want to be extra safe, you could also escape the values on the db with a `before insert trigger`.

Let me know if that clears your doubts.

**Reply by monokrome** [reply]
Unless I misunderstand what you mean by client here, I want to chime in to remind people that doing anything like this client-side is **never** good advice, because you can't guarantee that an exploited client is not being used to send malformed data.

As an example, if I create a web client in Python that acts like it's your JavaScript page and authenticates a user or otherwise gets a token, I can avoid any client-side filter you may have set up and inject whatever data that I want if there isn't protection on the server.

If you mean in the database client app (which I assumed isn't the case due to React being an example) then I apologize for misunderstanding the suggestion but still think it's a reasonable clarification to make.

**Reply by steve-chavez** [reply]
Thanks for chiming in @monokrome.

I meant web clients. So I'm not sure what OP really meant with *sanitizing*, but here you refer to the XSS part.

> I can avoid any client-side filter you may have set up and inject whatever data that I want if there isn't protection on the server.

Yes, but it's the job of the web app to escape output to render safe HTML content. If the attacker can impersonate the web app(inject his own js+html), then we'd be talking about MITM attacks - and HTTPS should be able to neutralize these.

That being said, even js frameworks that ensure escaping could have their leaks(React has [dangerouslySetInnerHTML](https://zhenyong.github.io/react/tips/dangerously-set-inner-html.html)) if the developer is not careful enough. So it'd definitely be better if we could ensure safe html with a correct type on the db.

I think a domain would do best here(better than my previous trigger suggestion). Like:

```sql
create domain alphanumeric AS text CHECK (value ~ '^[A-Z0-9]+$');
```

This would quickly reject any malicious values like `<script>`, of course this domain depends on the use case.
Having many of these domains could reduce write throughput, but if you want to be extra safe it's worth it.

WDYT?

**Reply by monokrome** [reply]
Oh, I thought you meant for escaping data before putting it in the database - maybe this is not as big an issue in this case then. Domain seems pretty good, though! Thanks for the response 😺
